addOns/selenium/src/main/java/org/zaproxy/zap/extension/selenium/PopupMenuItemOpenCustomRequestInBrowser.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2022 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.extension.selenium;
+
+import org.apache.commons.httpclient.URIException;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.parosproxy.paros.network.HttpMessage;
+import org.parosproxy.paros.view.View;
+import org.zaproxy.zap.extension.api.API;
+
+@SuppressWarnings("serial")
+public class PopupMenuItemOpenCustomRequestInBrowser extends PopupMenuItemOpenInBrowser {
+
+    private static final long serialVersionUID = 1L;
+
+    private static final Logger LOGGER =
+            LogManager.getLogger(PopupMenuItemOpenCustomRequestInBrowser.class);
+
+    private ExtensionSelenium ext;
+
+    private ProvidedBrowser browser;
+
+    public PopupMenuItemOpenCustomRequestInBrowser(
+            String name, ExtensionSelenium ext, ProvidedBrowser browser) {
+        super(name, ext, browser);
+        this.ext = ext;
+        this.browser = browser;
+    }
+
+    @Override
+    public void performAction(HttpMessage msg) {
+        new Thread(
+                        () -> {
+                            this.openInBrowser(msg);
+                        })
+                .start();
+    }
+
+    void openInBrowser(HttpMessage msg) {
+        try {
+            int id = msg.getHistoryRef().getHistoryId();
+            String url =
+                    API.getInstance()
+                            .getCallBackUrl(this.ext.getApiImplementor(), this.getSite(msg));
+            url = url + "?hist=" + id;
+            url = url + '&' + msg.getRequestHeader().getURI().getQuery();
+            ext.getProxiedBrowser(browser.getId(), url);
+        } catch (Exception e) {
+            View.getSingleton().showWarningDialog(e.getMessage());
```

Comment:
I seemed to hit the exception case more often than not.

I browsed to demo.testfire.net clicked a few pages, submitted the login POST with any old values. Then tried to open a few of the things via the new menu.

I don't really understand what's "Custom" about it, but simple GETs seemed to work while other GETs failed, and the one POST I had to work with never worked.

Reply:
Okay so for the POST requests, I think we need to follow the redirect/"follow the lead" to the very next GET request. Actually, I guess this was the same concern in the issue itself. I did need some views on what could be done. The main "concern" was to send modified headers and load the "result" in a browser to kind of verify/view the exploit/vulnerability.